### ISSUE TITLE
Can't resolve graphl file when using Webpack 5 (and nextjs)

### ISSUE THREAD

**Issue by outerlook** [facebook/relay#3272]
Hi.

Using:
- Next.js
- Babel relay plugin
- typescript (projects)
- Default webpack config (which is 4)
- Yarn workspaces
everything runs fine.

But when I upgrade to webpack 5, which is supported by nextjs I get the following error while starting development server or building:

`Module not found: Can't resolve '..\..\__generated__\relay\someMutation.graphql'`
this is the graphql file added by babel relay plugin.

I've been trying to debug and understand better babel's work but I confess I do not have much knowledge to debug it further by myself.

babel.config.json
```
{
  "presets": [
    "next/babel"
  ],
  "plugins": [
    [
      "relay",
      {
        "artifactDirectory": "./src/__generated__/relay/"
      }
    ],
    "@babel/plugin-transform-runtime",
    "babel-plugin-macros",
    [
      "styled-components",
      {
        "ssr": true,
        "displayName": true
      }
    ]
  ]
}
```
Note that I use babel plugin macros for tailwind's macro, not relay plugin macro. Although I tried using it and I got the same error too...

My generated files are indeed correctly generated at `'.src/__generated__/relay'`

It does work with webpack 4, and not with webpack 5.

I appreciate any help to try debugging it, I didn't find any public repo that uses typescript, relay and webpack 5, that's why I'm not sure I'm the one doing something wrong.

**Comment by outerlook** [comment]
strangely, a workaround is to remove artifactDirectory from both react-compiler options and babel-relay-plugin... then it generates a __generated__ folder on each file that uses a graphql tag.
It messes up my codebase but it works I discover how to get it to work like before on webpack 5

**Comment by CoericK** [comment]
Hey @outerlook, did you manage to solve? I do have the same stack, and noticed that it only happends on Windows environments on Unix it works just fine. I will try the work around you suggested

**Comment by vgedzius** [comment]
The real culprit seems to be here: https://github.com/relay-tools/relay-compiler-language-typescript#typescript

Relay compiler requires module to be set to "es2015" in compilerOptions of tsconfig, however next automatically, sets it to "esnext". I am yet to find a workaround.

**Comment by farshid1** [comment]
@vgedzius any luck??

**Comment by CrocoDillon** [comment]
I run into the same issue:

```
Module not found: Can't resolve '..\..\.relay\someQuery.graphql'
Did you mean './..\..\.relay\someQuery.graphql'?
Requests that should resolve in the current directory need to start with './'.
```

Not sure where that error is coming from but at least I found the reason why the path doesn't start with `'./'`:

https://github.com/facebook/relay/blob/521ca914ec756d8c8d6f291dde93c3eccf1ae56f/packages/babel-plugin-relay/compileGraphQLTag.js#L236-L237

Apparently that was fine in Webpack 4 but not in Webpack 5? If the error comes from Webpack that is.

Changing that line to include `'./'` fixed the issue for me... at least as a temporary workaround.

**Comment by artola** [comment]
I confirm that for us also in Windows we see this error (except with `"module": "commonjs"` where it works).

It seems related to the `joinPath` (alias for `path.join`) because it is environment sensitive.

If we "patch" the code (by replacing the slashes) in the following line to make it POSIX ... then it works everywhere, independently of the settings in `tsconfig`.

https://github.com/facebook/relay/blob/ac930cd215152057114933c20d5fe1b02e2bc3bb/packages/babel-plugin-relay/compileGraphQLTag.js#L239

```js
return relativeReference + joinPath(relative, fileToRequire).replace(/\\/g, '/');
```
@alloy @kassens  Do you have some idea to define if it a problem of this plugin or next in the chain? (for example some change in webpack API between v4 and v5 regarding normalisation of the paths)

See: Tobias' comment https://blog.johnnyreilly.com/2021/04/20/ts-loader-goes-webpack-5/

> Note that asset names should have only / as they are filenames and not paths. Only absolute paths have \.

**Comment by sokra** [comment]
`getRelativeImportPath` is incorrect for windows. Requests in imports should always use `/` but path.relative returns a path which contains `\`.

**Comment by artola** [comment]
@sokra Do you understand that this should be fixed in the relay plugin? Why before in **v4** this code that it is 2 years old was passing?

**Comment by artola** [comment]
@jstejada @josephsavona I am of the same opinion than @sokra above and I caught in this because our Windows users started to complain while other users on mac or server in linux has not detected any problem.
As a fast solution, I wrote a small Babel plugin to fix the `require` only for `process.platform === 'win32'` running just after `babel-plugin-relay` and since then no problems anymore, so happy `webpack` and `jest` in any environment.

**Advise**: if you want to use `path.relative` but normalise it with  `.replace(/\\/g, '/');` . Please let me know if you prefer a PR.

**Comment by artola** [comment]
I someone experience this problem, might be interested to use this plugin that I did, just add it to your `babel.config.js` after the `relay` plugin. It is in the meantime, until this problem is fixed and a new `babel-plugin-relay` is published.

```js
...
plugins: [
  ['babel-plugin-relay', { ... }],
  [require.resolve('./pathNormaliser')],
  ...
],
...
```

```js
'use strict';

const replacer = (filepath) => filepath.replace(/\\/g, '/');

/**
 * Normalise path to POSIX (only for `win32` using `require`).
 */
const PathNormaliserPlugin = function (api) {
  if (process.platform === 'win32') {
    const t = api.types;

    return {
      name: 'path-normaliser',

      visitor: {
        CallExpression: {
          enter: function (nodePath) {
            const callee = nodePath.get('callee');

            if (callee.isIdentifier() && callee.equals('name', 'require')) {
              const arg = nodePath.get('arguments.0');

              if (arg && arg.isStringLiteral()) {
                const sourcePath = arg.node.value;
                const targetPath = replacer(sourcePath);

                if (sourcePath !== targetPath) {
                  arg.replaceWith(t.stringLiteral(targetPath));
                }
              }
            }
          },
        },
      },
    };
  }

  return {
    name: 'path-normaliser',
  };
};

module.exports = PathNormaliserPlugin;
```

**Comment by artola** [comment]
@alunyov Could you please advice how to proceed? is FB team fix ongoing? or should we PR?

**Comment by artola** [comment]
@alunyov This happens even with v13 since the code has not changed in the involved lines mentioned above:

https://github.com/facebook/relay/blame/b5add1e0369ee38150d810c5c677c57d0952670e/packages/babel-plugin-relay/compileGraphQLTag.js

I do appreciate the advise of @sokra that confirms the bug:

https://github.com/facebook/relay/issues/3272#issuecomment-912781359

This is just affecting non posix (Windows) and the required fix it is just a few characters long, should I create a PR for this or are you thinking to apply a different solution?

**Comment by wKich** [comment]
@artola I like that way. What's a problem with using this solution?

**Comment by artola** [comment]
> Fix getting relative import path in windows for babel-plugin-relay #3830

@wKich Thanks for the PR. From my point of view, this is the simplest solution and also Windows' compatible.

**Comment by Phat-Loc** [comment]
Does any have a repo working that uses Relay 12, Webpack 5 and typescript? There is a bunch of module resolution errors that I am unable resolve. Thanks!

**Comment by giautm** [comment]
Hey, if you're stuck with this issue, please try to use [patch-package](https://www.npmjs.com/package/patch-package) to apply the fix.

```diff
diff --git a/node_modules/babel-plugin-relay/lib/compileGraphQLTag.js b/node_modules/babel-plugin-relay/lib/compileGraphQLTag.js
index 39a21e6..928c8de 100644
--- a/node_modules/babel-plugin-relay/lib/compileGraphQLTag.js
+++ b/node_modules/babel-plugin-relay/lib/compileGraphQLTag.js
@@ -14,7 +14,7 @@ var crypto = require('crypto');
 var _require = require('graphql'),
     print = _require.print;
 
-var _require2 = require('path'),
+var _require2 = require('path/posix'),
     dirname = _require2.dirname,
     joinPath = _require2.join,
     relativePath = _require2.relative,
```

**Comment by wKich** [comment]
@artola I checked usage of `require('path').posix` and it doesn't fix backslashes in my case. So I think `replace(/\\/g, '/')` is the best solution, but we need to apply it only for windows

**Comment by artola** [comment]
> @artola I checked usage of `require('path').posix` and it doesn't fix backslashes in my case. So I think `replace(/\\/g, '/')` is the best solution, but we need to apply it only for windows

@giautm could you please confirm if your path working on Windows as expected?

**Comment by artola** [comment]
The above plugin is a workaround; a proper fix is on the way with PR #4544.
